packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1011,12 +1011,21 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     return _getTextPositionVertical(offset, verticalOffset);
   }
 
-  // Deletes the current uncollapsed selection.
-  void _deleteSelection(TextSelection selection, SelectionChangedCause cause) {
-    assert(selection.isCollapsed == false);
+  // Deletes the current non-empty selection.
+  //
+  // Operates on the text/selection contained in textSelectionDelegate, and does
+  // not depend on `RenderEditable.selection`.
+  //
+  // If the selection is currently non-empty, this method deletes the selected
+  // text and returns true. Otherwise this method does nothing and returns
+  // false.
+  bool _deleteNonEmptySelection(SelectionChangedCause cause) {
+    assert(!readOnly);
+    final TextSelection selection = textSelectionDelegate.textEditingValue.selection;
```

Comment:
We have used the `textSelectionDelegate.textEditingValue` in a lot of places.
maybe we should define a getter for textEditingValue?

Reply:
I'm thinking we may want to get rid of it in `RenderEditable`: https://github.com/flutter/flutter/issues/79420#issuecomment-837456362